python/pyarrow/_parquet.pyx

```diff
@@ -342,13 +343,13 @@ cdef logical_type_name_from_enum(ParquetLogicalType type_):
 cdef class ParquetReader:
     cdef:
         object source
-        MemoryPool* allocator
+        CMemoryPool* allocator
         unique_ptr[FileReader] reader
         column_idx_map
         FileMetaData _metadata
 
-    def __cinit__(self):
-        self.allocator = default_memory_pool()
+    def __cinit__(self, MemoryPool memory_pool=None):
```

Comment:
Those are two rather interesting things, both also need the C++ equivalent as it would be there as important as in Python. The subpools are definitely useful but could be implemented as simple proxy objects whereas the question of keeping MemoryPools alive is quite difficult. Currently we have only weak references to the Pool and let the user take care of the lifetime, either we make the references in the PoolBuffer to the MemoryPool a shared reference or we could simply add a `DCHECK(allocted_bytes == 0)` to the destructor of a Pool .

Rather than also doing them here (and in the 0.2 release) I would open respective JIRAs and merge this patch. Would that be ok for you?